src/sys/pthread.rs

```diff
@@ -34,10 +39,351 @@ pub fn pthread_kill<T>(thread: Pthread, signal: T) -> Result<()>
+///         sleep(DELTA);
+///         mutex.lock()?;
+///         assert!(now.elapsed() > TIMEOUT && now.elapsed() < TIMEOUT + DELTA);
+///     }
+/// }
+/// 
+/// # Ok(())
+/// # }
+/// ```
+#[cfg(target_os = "linux")]
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct Mutex(UnsafeCell<libc::pthread_mutex_t>);
+
+#[cfg(target_os = "linux")]
+impl Mutex {
+    /// Wraps [`libc::pthread_mutex_init`].
+    /// 
+    /// # Safety
+    /// 
+    /// Attempting to initialize an already initialized mutex results in undefined behavior.
+    pub unsafe fn init(mutex: *mut Mutex, attr: Option<MutexAttr>) -> Result<()> {
+        let attr = match attr {
+            Some(mut x) => &mut x.0,
+            None => std::ptr::null_mut()
+        };
+        Errno::result(libc::pthread_mutex_init((*mutex).0.get(),attr))?;
+        Ok(())
+    }
+    /// Wraps [`libc::pthread_mutex_init`].
+    pub fn new(attr: Option<MutexAttr>) -> Result<Self> {
+        let attr = match attr {
+            Some(mut x) => &mut x.0,
+            None => std::ptr::null_mut()
+        };
+        let init = unsafe {
+            let mut uninit = std::mem::MaybeUninit::<libc::pthread_mutex_t>::uninit();
+            Errno::result(libc::pthread_mutex_init(uninit.as_mut_ptr(),attr))?;
+            uninit.assume_init()
+        };
+        Ok(Self(UnsafeCell::new(init)))
+    }
+    /// Wraps [`libc::pthread_mutex_lock`].
+    /// 
+    /// <https://man7.org/linux/man-pages/man3/pthread_mutex_lock.3p.html>
+    pub fn lock(&self) -> Result<MutexGuard<'_>> {
+        unsafe {
+            Errno::result(libc::pthread_mutex_lock(self.0.get())).map(|_| MutexGuard(self))
+        }
+    }
+    /// Wraps [`libc::pthread_mutex_trylock`].
+    /// 
+    /// <https://man7.org/linux/man-pages/man3/pthread_mutex_lock.3p.html>
+    pub fn try_lock(&self) -> Result<Option<MutexGuard<'_>>> {
+        unsafe {
+            match Errno::result(libc::pthread_mutex_trylock(self.0.get())) {
+                Ok(_) => Ok(Some(MutexGuard(self))),
+                Err(Errno::EBUSY) => Ok(None),
+                Err(err) => Err(err)
+            }
+            
+        }
+    }
+    /// Wraps [`libc::pthread_mutex_unlock`].
+    /// 
+    /// <https://man7.org/linux/man-pages/man3/pthread_mutex_lock.3p.html>
+    /// 
+    /// Prefer unlocking by dropping the [`MutexGuard`] returned by [`Mutex::lock`] or [`Mutex::try_lock`].
+    /// 
+    /// # Safety
+    /// 
+    /// Results in UB if not called from the thread that locked the mutex.
+    pub unsafe fn unlock(&self) -> Result<()> {
+        unsafe {
+            Errno::result(libc::pthread_mutex_unlock(self.0.get())).map(drop)
+        }
+    }
+}
+
+#[cfg(target_os = "linux")]
+unsafe impl Sync for Mutex {}
+
+#[cfg(target_os = "linux")]
+impl std::ops::Drop for Mutex {
+    /// Wraps [`libc::pthread_mutex_destroy`].
+    fn drop(&mut self) {
+        unsafe {
+            Errno::result(libc::pthread_mutex_destroy(self.0.get())).unwrap();
```

Comment:
Panicing during drop is very bad.  It makes bugs hard to debug.  So you should check if the thread is already panicing, and if so don't unwrap the Result.

Reply:
I'm not sure I entirely agree with this approach as it adds unavoidable overhead. Is there some more documentation on why to use this approach?

Reply:
Panicing while already unwinding a panic causes a SIGILL signal.  So rather than seeing a nice stack trace, the user just sees that his program was killed by SIGILL.

Reply:
To an end user (not the developer) I don't think they really care about debug info. I generally take the approach any panic is disastrous. Maybe it'd make sense to do something like:
```rust
impl std::ops::Drop for MutexAttr {
    /// Wraps [`libc::pthread_mutexattr_destroy`].
    fn drop(&mut self) {
        let res = libc::pthread_mutexattr_destroy(&mut self.0);
        #[cfg(debug_assertions)]
        if std::thread::panicked() {
            return;
        }
        Errno::result(res).unwrap();
    }
}
```